common/src/service/ExceptionHandlingHelper.cs

```diff
@@ -36,5 +60,70 @@ public static Task<string> GetExceptionMessageAsync(HttpResponseMessage response
+            // the error description in the response header. Therefore, there is a chance that the 6-digit error code does not match the error description. For that reason,
+            // the SDK will do its best to decide what to surface to the user.
+            // The SDK will attempt to retrieve the integer error code from the response content and the error description from the response header. Through a 'description'
+            // to 'error code' enum mapping, the SDK will check if both values are a match. If so, the SDK will populate the exception with the proper Code. In the case where
+            // there is a mismatch between the error code and the description, the SDK returns ErrorCode.InvalidErrorCode and log a warning. 
+
+            int errorCode = -1;
+            try
+            {
+                IoTHubExceptionResult responseContent = JsonConvert.DeserializeObject<IoTHubExceptionResult>(responseContentStr);
+                Dictionary<string, string> messageFields = JsonConvert.DeserializeObject<Dictionary<string, string>>(responseContent._message);
+                if (messageFields != null && messageFields.TryGetValue(CommonConstants.ErrorCode, out string errorCodeObj))
+                {
+                    errorCode = Convert.ToInt32(errorCodeObj, CultureInfo.InvariantCulture);
+                }
+                else
+                {
+                    return ErrorCode.InvalidErrorCode;
+                }
+            }
+            catch (JsonReaderException ex)
+            {
+                if (Logging.IsEnabled)
+                {
+                    Logging.Error(null, $"Failed to parse response content JSON: {ex}. Message body: '{responseContentStr}.'");
+                }
+
+                return ErrorCode.InvalidErrorCode;
+            }
+
+            // Now that we retrieved the integer error code from the response content, we will retrieve the error description from the header.
+            string headerErrorCodeString = response.Headers.GetFirstValueOrNull(CommonConstants.HttpErrorCodeName);
+            if (headerErrorCodeString != null &&
+                Enum.TryParse(headerErrorCodeString, out ErrorCode headerErrorCode))
+            {
+                if ((int)headerErrorCode == errorCode)
+                {
+                    // We have a match. Therefore, return the proper error code.
+                    return headerErrorCode;
+                }
+                else
+                {
+                    if (Logging.IsEnabled)
+                    {
+                        Logging.Error(null, $"There is a mismatch between the error code retrieved from the response content and the response header." +
+                            $"Content error code: {errorCode}. Header error code description: {(int)headerErrorCode}.");
+                    }
+                }
+            }
+
+            return ErrorCode.InvalidErrorCode;
```

Comment:
This is a lot of code to maintain in the SDK. This is great for now but we should probably let the service know to send back the right kind of response. This might be something they can fix as a non-breaking change.

Reply:
Agreed. The integer ErrorCode and description should be coming from a single place, as opposed to two different places. Do you have any contact in IoT Hub who can help clarify this?

Reply:
I don't. But we should ask Priyanka/Sejal and just send them an email.

Reply:
Agree with you guys here, we can put this workaround for now, but you should follow up with Srivani about this.